src/halmos/utils.py

```diff
@@ -349,6 +353,14 @@ def int_of(x: Any, err: str = "expected concrete value but got") -> int:
     if isinstance(res, int):
         return res
 
+    # if the value is still symbolic at this point, check if it can be evaluated
+    # concretely with the current solver state
+    if solver is not None and solver.check() == sat:
+        m = solver.model()
+        res = m.evaluate(x)
+        if is_bv_value(res):
+            return res.as_long()
+
```

Comment:
could you elaborate on the context of this?

Reply:
see the explanation in sha3_expr:

            # we have a choice here: return the concrete hash value or the symbol
            # - if we return the value, we can directly process expressions
            # - if we return the symbol, we can use the extra information to help
            #   with things like recovering storage base slots
            #
            # for now, we return the symbol, and if we need a concrete value, we can
            # lazily evaluate it at the point of use
            #
            # see: halmos.utils:int_of

returning the concrete hash makes evaluating expressions easy, but breaks things that expect to parse symbols like the storage processing code. Returning symbols like `f_sha3_N(x)` keeps the storage processing code happy, but breaks if you want to do something like `mload(1 + f_sha3_N(x))`

so the middle ground I went with is to keep the expression symbolic, but when we attempt to get a concrete value out of it, we need a way to replace `f_sha3_256(0)` with the actual value. Using `model.evaluate` seems to do the trick efficiently

Reply:
the problem is that we need a solver to do the evaluation (instead of just erroring out), and so I shoved one in `State` which is arguably not great.

Reply:
specifically it lets us support `check_concrete_keccak_memory_lookup`

Reply:
ok it is indeed not sound:

```
>>> s = Solver()
>>> s.add(x < 3)
>>> s.check()
sat
>>> m = s.model()
>>> m.evaluate(x)
2
>>> m.evaluate(x + 1)
3
```

looks like we need to substitute the symbol with its value then